### PR TITLE
Role Types No Longer Types

### PR DESCRIPTION
**Pull request by fppt**

### CODE REVIEW

File: grakn-core/src/main/java/ai/grakn/concept/OntologyElement.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+ *
+ *
+ * @author fppt
+ */
+public interface OntologyElement extends Concept {
+    //------------------------------------- Accessors ---------------------------------
+    /**
+     * Returns the unique id of this Type.
+     *
+     * @return The unique id of this type
```
Comment (aelred) [inline grakn-core/src/main/java/ai/grakn/concept/OntologyElement.java]: These docs still say "type".

File: grakn-core/src/main/java/ai/grakn/concept/TypeLabel.java [hunk 2]
```diff
@@ -24,7 +24,7 @@
 
 /**
  * <p>
- *     A Type Label
+ *     A Label
```
Comment (aelred) [inline grakn-core/src/main/java/ai/grakn/concept/TypeLabel.java]: Perhaps should include "...for an ontology element"

File: grakn-core/src/main/java/ai/grakn/exception/GraphOperationException.java [hunk 3]
```diff
@@ -108,8 +109,8 @@ public static GraphOperationException duplicateHas(Type type, ResourceType resou
     /**
      * Thrown when setting {@code superType} as the super type of {@code type} and a loop is created
      */
-    public static GraphOperationException loopCreated(Type type, Type superType){
-        throw new GraphOperationException(ErrorMessage.SUPER_TYPE_LOOP_DETECTED.getMessage(type.getLabel(), superType.getLabel()));
+    public static GraphOperationException loopCreated(OntologyElement type, OntologyElement superElement){
```
Comment (aelred) [inline grakn-core/src/main/java/ai/grakn/exception/GraphOperationException.java]: Forgot to rename the first parameter

File: grakn-graph/src/main/java/ai/grakn/graph/internal/OntologyElementImpl.java [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+ *     These differ from normal graph constructs in two ways:
+ *     1. They have a unique {@link TypeLabel} which identifies them
+ *     2. You can link them together into a hierarchical structure
+ * </p>
+ *
+ * @author fppt
+ *
+ * @param <T> The leaf interface of the object concept.
+ *           For example an {@link EntityType} or {@link RelationType} or {@link RoleType}
+ */
```
Comment (aelred) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/OntologyElementImpl.java]: Is it worth repeating these docs for the implementation? To me it seems like one more place for the docs to be inconsistent.

File: grakn-graph/src/main/java/ai/grakn/graph/internal/OntologyElementImpl.java [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+public abstract class OntologyElementImpl<T extends OntologyElement> extends ConceptImpl implements OntologyElement {
+    private final TypeLabel cachedLabel;
+    private final TypeId cachedLabelId;
+
+    private Cache<T> cachedSuperType = new Cache<>(() -> this.<T>neighbours(Direction.OUT, Schema.EdgeLabel.SUB).findFirst().orElse(null));
+    private Cache<Set<T>> cachedDirectSubTypes = new Cache<>(() -> this.<T>neighbours(Direction.IN, Schema.EdgeLabel.SUB).collect(Collectors.toSet()));
```
Comment (aelred) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/OntologyElementImpl.java]: These probably shouldn't refer to "type" anymore either.

File: grakn-graph/src/main/java/ai/grakn/graph/internal/TypeImpl.java [hunk 6]
```diff
@@ -238,105 +232,22 @@ public TypeId getTypeId(){
...
-    }
-
-    /**
-     * Removes an old sub type from the currently cached sub types. If no subtypes have been cached then this will hit the database.
-     *
-     * @param oldSubType The old sub type which should not be cached anymore
-     */
-    private void deleteCachedDirectedSubType(T oldSubType){
-        cachedDirectSubTypes.ifPresent(set -> set.remove(oldSubType));
+        return Collections.unmodifiableCollection(filterImplicitStructures(super.subTypes()));
```
Comment (aelred) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/TypeImpl.java]: This behaviour should probably be in `OntologyElement`

**Comment by fppt** [comment]
!rtg - Failed tests: 
  BatchMutatorClientTest.whenSending50InsertQueries_50EntitiesAreLoadedIntoGraph:94 expected:<100> but was:<75>

File: grakn-core/src/main/java/ai/grakn/concept/Concept.java [hunk 7]
```diff
@@ -53,176 +53,192 @@
...
 
     /**
-     * Return as a Type if the Concept is a Type.
+     * Return as a {@link OntologyConcept} if the {@link Concept} is a {@link OntologyConcept}.
      *
-     * @return A Type if the concept is a Type
+     * @return A {@link OntologyConcept} if the {@link Concept} is a {@link OntologyConcept}
+     */
+    @CheckReturnValue
+    OntologyConcept asOntologyElement();
```
Comment (aelred) [inline grakn-core/src/main/java/ai/grakn/concept/Concept.java]: ought to be `asOntologyConcept`

File: grakn-core/src/main/java/ai/grakn/concept/Concept.java [hunk 8]
```diff
@@ -53,176 +53,192 @@
...
     Rule asRule();
 
     /**
-     * Determine if the Concept is a Type.
+     * Determine if the {@link Concept} is a {@link OntologyConcept}
+     *
+     * @return true if the{@link Concept} concept is a {@link OntologyConcept}
+     */
+    @CheckReturnValue
+    boolean isOntologyElement();
```
Comment (aelred) [inline grakn-core/src/main/java/ai/grakn/concept/Concept.java]: `isOntologyConcept`

File: grakn-graph/src/main/java/ai/grakn/graph/internal/OntologyConceptImpl.java [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+    }
+
+    /**
+     * Deletes the concept as an Ontology Element
+     */
+    @Override
+    public void delete(){
+        checkOntologyMutationAllowed();
+        boolean hasSubs = neighbours(Direction.IN, Schema.EdgeLabel.SUB).findAny().isPresent();
+        boolean hasInstances = currentShard().links().findAny().isPresent();
```
Comment (aelred) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/OntologyConceptImpl.java]: Arguably this check should be in `TypeImpl`, but it does no harm here

File: grakn-graph/src/main/java/ai/grakn/graph/internal/TypeImpl.java (hunk 6 above)
Comment (aelred) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/TypeImpl.java]: You can get rid of this override entirely now 👍

File: grakn-graph/src/test/java/ai/grakn/generator/GraknGraphs.java [hunk 10]
```diff
@@ -312,10 +318,16 @@ private TypeLabel typeLabel() {
         return gen().make(TypeLabels.class, gen().make(MetasyntacticStrings.class)).generate(random, status);
     }
 
-    private Type type() {
+    private OntologyConcept ontologyElement() {
```
Comment (aelred) [inline grakn-graph/src/test/java/ai/grakn/generator/GraknGraphs.java]: Method should be `ontologyConcept`

File: grakn-graph/src/test/java/ai/grakn/generator/GraknGraphs.java [hunk 11]
```diff
@@ -337,7 +349,11 @@ private RuleType ruleType() {
     }
 
     private Thing instance() {
-        return chooseOrThrow(graph.admin().getMetaConcept().instances());
+        Set<? extends Thing> candidates = graph.admin().getMetaConcept().subTypes().stream().
+                filter(element -> !element.isRoleType()).
```
Comment (aelred) [inline grakn-graph/src/test/java/ai/grakn/generator/GraknGraphs.java]: Yeah, `role` shouldn't sub `thing` :P

**Pull request merged**